File: LeetCode/All/js/61.rotate-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var rotateRight = function(head, k) {
    const listLength = findListLength(head);
    const intersectionPoint = k%listLength;

    if(listLength <= 1) return head;
    if(listLength === intersectionPoint || intersectionPoint === 0) return head;
    let tempHead = head;
    let i = 0;
    while(i < (listLength - intersectionPoint - 1)){
        tempHead = tempHead.next;
        i++;
    }

    let newHead = tempHead.next;
    let newTemp = tempHead.next;
    tempHead.next = null;

    while(newTemp.next!=null){
        newTemp = newTemp.next;
    }

    newTemp.next = head;
    head = newHead;

    return head;

};

const findListLength = (head) => {
    let length = 0;
    while(head!=null){
        length++;
        head = head.next;
    }
    return length;
}

File: LeetCode/Tag/Two-Pointers/cpp/61.rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
    int getListLength(ListNode* head) {
        ListNode *curr = head;
        int len = 0;
        while (curr) {
            ++len;
            curr = curr->next;
        }
        return len;
    }

public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (!head) return NULL;
        k %= getListLength(head);
        if (k == 0) return head;
        
        ListNode *p1 = head, *p2 = head;
        for (int i = 0; i < k; ++i) {
            p2 = p2->next;
        }
        while (p2->next) {
            p1 = p1->next;
            p2 = p2->next;
        }
        p2->next = head;
        head = p1->next;
        p1->next = NULL;
        return head;
    }
};